File: vehicle-tracker-backend/identity-service/src/main/java/com/github/diogocerqueiralima/identity/service/infrastructure/mappers/DeviceMapper.java
package com.github.diogocerqueiralima.identity.service.infrastructure.mappers;

import com.github.diogocerqueiralima.identity.service.domain.model.device.Device;
import com.github.diogocerqueiralima.schema.proto.DeviceResponse;

import java.util.UUID;

/**
 * This class is responsible for mapping between Device domain objects and their corresponding gRPC responses.
 */
public class DeviceMapper {

    // Should not be instantiated
    private DeviceMapper() {}

    /**
     *
     * Converts a {@link DeviceResponse} object to a {@link Device} domain object.
     *
     * @param response the {@link DeviceResponse} object to be converted
     * @return a {@link Device} domain object corresponding to the provided response
     */
    public static Device toDomain(DeviceResponse response) {
        return new Device(
                UUID.fromString(response.getId()),
                UUID.fromString(response.getOwnerId())
        );
    }

}

File: vehicle-tracker-mobile/app/src/main/java/com/github/diogocerqueiralima/infrastructure/devices/mappers/DeviceMapper.kt
package com.github.diogocerqueiralima.infrastructure.devices.mappers

import com.github.diogocerqueiralima.domain.devices.model.Device
import com.github.diogocerqueiralima.infrastructure.devices.entities.DeviceEntity
import java.util.UUID

/**
 * Extension function to convert a [DeviceEntity] infrastructure entity to a [Device] domain model.
 */
fun DeviceEntity.toDomain() = Device(
    id = UUID.fromString(this.id),
    createdAt = this.createdAt,
    updatedAt = this.updatedAt,
    serialNumber = this.serialNumber,
    model = this.model,
    manufacturer = this.manufacturer,
    imei = this.imei
)
